src/mlir_egglog/mlir_gen.py

```diff
@@ -191,81 +194,77 @@ def lookup(e):
         self.cache[expr] = as_source(expr, self.vars, lookup)
 
 
-def get_children(expr: ir.Expr) -> set[ir.Expr]:
+def get_children(expr: ir.Term) -> set[ir.Term]:
     """Get child expressions for an AST node."""
-    match expr:
-        case ir.BinaryOp():
-            return {expr.lhs, expr.rhs}
-        case ir.UnaryOp():
-            return {expr.operand}
-        case ir.FloatLiteral() | ir.IntLiteral() | ir.Symbol():
-            return set()
-        case _:
-            raise NotImplementedError(
-                f"Cannot get children for expression type: {type(expr)}\n"
-                f"Expression: {expr}\n"
-                f"Supported types: BinaryOp (Add, Mul, Div, Maximum), UnaryOp (Sin, Cos, etc.), "
-                f"FloatLiteral, IntLiteral, Symbol\n"
-                f"This error typically occurs when trying to process an unsupported operation."
-            )
+    return {child for child in get_callable_args(expr) if isinstance(child, ir.Term)}
 
 
 def as_source(
-    expr: ir.Expr, vars: list[str], lookup_fn: Callable[[ir.Expr], str]
+    expr: ir.Term, vars: list[str], lookup_fn: Callable[[ir.Term], str]
 ) -> str:
     """
     Convert expressions to MLIR source code using arith and math dialects.
     """
-    match expr:
-        # Literals and Symbols
-        case ir.FloatLiteral(fval=val):
-            return f"arith.constant {val:e} : {F32_TYPE}"
-        case ir.IntLiteral(ival=val):
-            return f"arith.constant {val} : {I32_TYPE}"
-        case ir.Symbol(name=name) if name in vars:
-            return f"%arg_{name}"
-        case ir.Symbol(name=name):
-            return f"%{name}"
-
-        # Binary Operations
-        case ir.Add(lhs=lhs, rhs=rhs):
+    # Literals and Symbols
+    match get_callable_args(expr, ir.Term.lit_f32):
+        case (f64(f),):
+            return f"arith.constant {f:e} : {F32_TYPE}"
+    match get_callable_args(expr, ir.Term.lit_i64):
+        case (i64(i),):
+            return f"arith.constant {i} : {I32_TYPE}"
```

Comment:
This is a bit off that the "i64" is emitted as an "i32" but I think this preserves the behavior that you had previously.